src/benchmarks/pmemobj_tx_add_range.cpp

```diff
@@ -0,0 +1,241 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *      * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *      * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *      * Neither the name of the copyright holder nor the names of its
+ *        contributors may be used to endorse or promote products derived
+ *        from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * pmemobj_tx_add_range.cpp -- pmemobj_tx_add_range benchmarks definition
+ */
+
+#include <cassert>
+#include <cerrno>
+#include <cstddef>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <fcntl.h>
+#include <sys/file.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include "benchmark.hpp"
+#include "file.h"
+#include "libpmemobj.h"
+
+#define LAYOUT_NAME "tx_add_range_benchmark"
+
+/*
+ * RANGE_COEFFICIENT -- pool has to hold every allocated object with
+ * its snapshot (1 + 1), plus 0.5 because of fragmentation
+ */
+#define RANGE_COEFFICIENT (1 + 1 + 0.5)
+
+/*
+ * obj_bench_args -- benchmark specific command line options
+ */
+struct obj_bench_args {
+	uint64_t nranged; /* number of allocated objects */
+};
+
+/*
+ * obj_bench -- benchmark context
+ */
+struct obj_bench {
+	PMEMobjpool *pop; /* persistent pool handle */
+	PMEMoid *ranged;  /* array of allocated objects */
+	size_t obj_size;  /* size of allocated object */
+	uint64_t nranged; /* number of allocated objects */
+};
+
+/*
+ * init_objects -- allocate persistent objects
+ */
+static int
+init_objects(struct obj_bench *ob)
+{
+	assert(ob->nranged != 0);
+	ob->ranged = (PMEMoid *)malloc(ob->nranged * sizeof(PMEMoid));
+
+	if (!ob->ranged) {
+		perror("malloc");
+		return -1;
+	}
+
+	for (uint64_t i = 0; i < ob->nranged; i++) {
+		PMEMoid *oid = &ob->ranged[i];
+		if (pmemobj_alloc(ob->pop, oid, ob->obj_size, 0, nullptr,
```

Comment:
Not sure if 1 object == 1 snapshot is a good idea, especially if we were to test very small sizes (min allocation is 128 bytes).